modules/layers/src/text-layer/font-atlas-manager.ts

```diff
@@ -188,26 +197,98 @@ export default class FontAtlasManager {
     }
 
     // update fontAtlas with new settings
-    const fontAtlas = this._generateFontAtlas(this._key, charSet, cachedFontAtlas);
+    const fontAtlas = this._generateFontAtlas(charSet, cachedFontAtlas);
     this._atlas = fontAtlas;
 
     // update cache
     cache.set(this._key, fontAtlas);
   }
 
-  private _generateFontAtlas(
-    key: string,
-    characterSet: Set<string>,
-    cachedFontAtlas?: FontAtlas
-  ): FontAtlas {
-    const {fontFamily, fontWeight, fontSize, buffer, sdf, radius, cutoff} = this.props;
+  private _generateFontAtlas(characterSet: Set<string>, cachedFontAtlas?: FontAtlas): FontAtlas {
     let canvas = cachedFontAtlas && cachedFontAtlas.data;
     if (!canvas) {
       canvas = document.createElement('canvas');
       canvas.width = MAX_CANVAS_WIDTH;
     }
+
     const ctx = canvas.getContext('2d')!;
 
+    return this.props.sdf
+      ? this._generateSDFFontAtlas(canvas, ctx, characterSet)
+      : this._generateRegularFontAtlas(canvas, ctx, characterSet, cachedFontAtlas);
+  }
+
+  private _generateSDFFontAtlas(
+    canvas: HTMLCanvasElement,
+    ctx: CanvasRenderingContext2D,
+    characterSet: Set<string>
+  ): FontAtlas {
+    const {fontFamily, fontWeight, fontSize, buffer, radius, cutoff} = this.props;
+    const mapping: CharacterMapping = {};
+    const tinySDF = new TinySDF({
+      fontSize,
+      buffer,
+      radius,
+      cutoff,
+      fontFamily,
+      fontWeight: `${fontWeight}`
+    });
+    const cellSize = fontSize + buffer * 2;
+    const layoutRowHeight = fontSize * HEIGHT_SCALE;
+    const columns = Math.floor(canvas.width / cellSize);
+    const rows = Math.ceil(characterSet.size / columns);
+    const requiredCanvasHeight = nextPowOfTwo(rows * cellSize);
+
+    if (canvas.height !== requiredCanvasHeight) {
+      resizeCanvas(ctx, canvas.width, requiredCanvasHeight);
+    }
+
+    let x = 0;
+    let y = 0;
+    for (const char of characterSet) {
```

Comment:
You are basically duplicating the `buildMappings` function to handle dynamic character width returned by `SDF.draw`. What advantages does this provide? I see that this new code still assumes that `width` and `height` returned by `tinySDF.draw` is contained in the `cellSize` square, is that true? If so, can we modify `buildMapping` to always use `cellSize` as the packing width in SDF mode?